### ISSUE TITLE
ENH: Overhaul of NumPy `lib` namespace [NEP 52]

### ISSUE THREAD

**Issue by mtsokol** [numpy/numpy#24507]
### Proposed new feature or change:

Hi @rgommers @ngoldbaum,

This issue is meant for monitoring and discussing cleanup of `np.lib` namespace. (Tracking parent issue: https://github.com/numpy/numpy/issues/23999)

Main goal is to make each public function, which lives in `np.lib`, available either from the main namespace `np` or "local lib" namespace e.g. `np.lib.npyio`. As a result, there will be only one way to access a specific public function in NumPy. 
To achieve it, for each relevant `lib` submodule, there will be a private file (that starts with `_`) imported by the main namespace, and a public file that allows to be accessed as a public submodule.

```python
# Example:

np.broadcast_to  # imported from numpy/lib/_stride_tricks.py

np.lib.stride_tricks.normalize_axis_tuple  # imported from numpy/lib/stride_tricks.py
```

Let's call this mechanism "main or submodule namespace access". (maybe a better name?)

---
### The scope of the milestone:
- [ ] introduce `np.lib.array_utils` submodule
- [ ] add `lib._histograms`
- [ ] add `lib._type_check`
- [ ] add `lib._nanfunctions`
- [ ] add `lib._function_base`
- [ ] add `lib._shape_base`
- [ ] add `lib._arraypad`
- [ ] add `lib._arraysetops`
- [ ] add `lib._twodim_base`
- [ ] add `lib._ufunclike`
- [ ] add `lib._utils`
- [ ] add `lib._npyio`
- [ ] add `lib._polynomial`
- [ ] add `lib._stride_tricks`
- [ ] add `lib._index_tricks`

**Comment by mhvk** [comment]
@mtsokol - be careful - `lib.stride_tricks` is meant to remain (I didn't check all others, but this is one I use (too?) regularly, and the contents are not all exported; see NEP).

**Comment by mhvk** [comment]
Yes, `as_strided` remains a favourite... There's also `sliding_window_view` (but `normalize_axis_tuple` comes from elsewhere).

**Comment by rgommers** [comment]
Focusing on what remains or becomes a public name would be useful. `stride_tricks` indeed, but there is more. `array_utils` is new. What about `index_tricks`, I think we were considering keeping that and exposing some of the more niche or advanced indexing related utilities there? 

The checklist of what to make private is useful too, but that's more a todo list. I'd suggest going into more detail on what will be officially public. And then on what the TBDs are or potential pain points. If some current `.py` file contains only objects that are in the main namespace already, then it's easy - that `.py` file does not need to be a public module. When that's not the case, we have to surface the changes.

**Comment by rgommers** [comment]
I am not sure, but I suspect that `normalize_axis_tuple` being in `stride_tricks` was an accident - it has nothing to do with strides, and I'd expect it in another place (`array_utils` maybe?).

**Comment by mhvk** [comment]
`normalize_axis_tuple` is imported from `numpy.core.numeric`, no worries about that (though there was a separate question of where it should be exposed - but not here for sure!)

**Comment by rgommers** [comment]
Thanks Mateusz, that's a clear list. There really is very little in "local ns", so basically we're hiding almost everything that's there now and have more or less a clean slate to add new niche functionality that doesn't qualify for inclusion in the main namespace. That'll be quite nice, and has been on our wish list for a long time.

Maybe add _local ns_ items for `numpy.lib` itself? I think that there's a couple of items in it, like `NumpyVersion` and `Arrayterator`.

**Comment by ngoldbaum** [comment]
It looks like there's still two minor usages of `import *` in `numpy/lib/__init__.py` that could be replaced but other than that this is done. Going to close this as I don't think those usages are bad, they're doing `import *` from modules with an `__all__`.
